ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,232 @@
+    // if count ever exceeds 2, we drop reports, but keep counting
+    // an honest client and report collector will only submit
+    // one report with a breakdown key and one report with a value.
+    // if there are less, it's unattributed. if more, something went wrong.
+    for report in reports {
+        let match_key = report.match_key;
+        let entry = reports_by_matchkey.entry(match_key).or_insert((
+            [
+                AggregateableHybridReport::<BK, V>::ZERO,
+                AggregateableHybridReport::<BK, V>::ZERO,
+            ],
+            0,
+        ));
+        if entry.1 == 0 {
+            // If the count is 0, replace the first element with the new report
+            entry.0[0] = report.into();
+            entry.1 += 1;
+        } else if entry.1 == 1 {
+            // If the count is 1, replace the second element with the new report
+            entry.0[1] = report.into();
+            entry.1 += 1;
+        } else {
+            // If the count is 2 or more, increment the counter and drop the report
+            entry.1 += 1;
+        }
+    }
+
+    // we only keep the reports from match_keys that provided exactly 2 reports
+    let report_pairs: Vec<[AggregateableHybridReport<BK, V>; 2]> = reports_by_matchkey
+        .into_iter()
+        .filter_map(|(_, v)| if v.1 == 2 { Some(v.0) } else { None })
+        .collect::<Vec<_>>();
+
+    let chunk_size = TARGET_PROOF_SIZE;
+    let ctx = ctx.set_total_records(TotalRecords::specified(report_pairs.len())?);
+
+    let dzkp_validator = ctx.dzkp_validator(
+        MaliciousProtocolSteps {
+            protocol: &HybridStep::GroupBySum,
+            validate: &HybridStep::GroupBySumValidate,
+        },
+        usize::MAX,
+    );
+
+    let agg_ctx = dzkp_validator.context();
+
+    let agg_work = stream::iter(report_pairs)
+        .enumerate()
+        .map(|(idx, reports)| {
+            let agg_ctx = agg_ctx.clone();
+            async move {
+                let (breakdown_key, _) = integer_add::<_, EightBitStep, 1>(
+                    agg_ctx.narrow(&AggregateReportsStep::AddBK),
+                    idx.into(),
+                    &reports[0].breakdown_key.to_bits(),
+                    &reports[1].breakdown_key.to_bits(),
+                )
+                .await?;
+                let (value, _) = integer_add::<_, EightBitStep, 1>(
+                    agg_ctx.narrow(&AggregateReportsStep::AddV),
+                    idx.into(),
+                    &reports[0].value.to_bits(),
+                    &reports[1].value.to_bits(),
+                )
+                .await?;
+                Ok::<_, Error>(AggregateableHybridReport::<BK, V> {
+                    match_key: (),
+                    breakdown_key: breakdown_key.collect_bits(),
+                    value: value.collect_bits(),
+                })
+            }
+        });
+
+    dzkp_validator
+        .validated_seq_join(agg_work)
```

Comment:
@andyleiserson This version works only with single batch and validated_seq_join. If number of batches is not `usize::MAX`, validation fails

Reply:
also if I use `validate_record` directly inside the closure instead of `validated_seq_join`

Reply:
This seems concerning, I will look at it.

Reply:
it may be HashMap but I don't think I fully understand why

Reply:
making this change:
```
let chunk_size: usize = TARGET_PROOF_SIZE / (BK::BITS as usize + V::BITS as usize);
```

and setting batch size to `chunk_size.next_power_of_two(),` seems to work!

Reply:
With or without your change, it only passes some of the time. With your change, it passes more often. Since the test only has two pairs that will actually be aggregated, it makes sense that it passes more often with more shards (less chance that both the pairs end up on the same shard). But I can't totally explain the pass rates that I am seeing. (I got 2 or 3 passes out of 40 before your change, 11 or 12 out of 40 after.)

Reply:
> making this change ... and setting batch size to chunk_size.next_power_of_two(), seems to work!

Are you sure it passes reliably? I really think the failures are due to `HashMap`.

 > it may be HashMap but I don't think I fully understand why

Considering a two party MPC for simplicity. If one helper has (share of A, share of B) and the other helper has (share of B, share of A), and they add those vectors together, the result is garbage. Taking proofs out of the picture, it would also fail a functional output check, the proofs just catch it sooner.

Reply:
@akoshelev and I just chatted about this live, and I think you're both right. the Hashmap will get things out of order, adn thus the proofs fail. I'm working on swapping out with a BTreeMap now.

Reply:
> Considering a two party MPC for simplicity. If one helper has (share of A, share of B) and the other helper has (share of B, share of A), and they add those vectors together, the result is garbage. Taking proofs out of the picture, it would also fail a functional output check, the proofs just catch it sooner.

yea that makes sense. I think I got tests failing in almost 100% of the cases **without** validated seq join and passing locally **with** validated seq join